### PR TITLE
Add tests to check the parsing of common AWS hostnames for region andservice type are correct; add support for ElasticSearch

### PR DESCRIPTION
**Pull request by matthew-andrews**

### CODE REVIEW

File: aws4.js [hunk 1]
```diff
@@ -27,6 +27,11 @@ function RequestSigner(request, credentials) {
   this.request = request
   this.credentials = credentials || this.defaultCredentials()
 
+  // ES's hostParts are the other way round
+  // E.g. search-cluster-name-aaaa00aaaa0aaa0aaaaaaa0aaa.us-east-1.es.amazonaws.com
```
Comment (mhart) [inline aws4.js]: I'm a little confused about this. There seem to be a couple of different hosts used by AWS ES:

http://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/what-is-amazon-elasticsearch-service.html#endpoints

That seems to suggest that the service name is actually `aes` when used with the `http://search-domainname-domainid.us-east-1.aes.amazonaws.com` type hostname, but `es` when used with a more standard `es.region.amazonaws.com` host name for configuration, etc.

Can you elaborate on this at all?

Comment (mhart) [inline aws4.js]: Same with here: http://docs.aws.amazon.com/general/latest/gr/rande.html#elasticsearch-service-regions

Are these configuration hosts, as opposed to service/data hosts? If so, should probably add a test to make sure they're supported as well (and that the hostnames aren't messed up)

Comment (matthew-andrews) [inline aws4.js]: Good spot, I'll add support for `es.region.amazonaws.com`

Comment (matthew-andrews) [inline aws4.js]: Was actually already supported (it only reversed if the value expected to be `region` was `es` — which can only happen in the `http://search-domainname-domainid.us-east-1.es.amazonaws.com` case…  But I've added an extra test :smile:

**Comment by mhart** [comment]
Cool! This looks great, thanks. Maybe just put the `reverse` logic back into `matchHost` – little bit of extra code, but that way it's always consistent in what it returns. So something like:

```js
RequestSigner.prototype.matchHost = function(host) {
  var match = (host || '').match(/([^\.]+)\.(?:([^\.]*)\.)?amazonaws\.com$/)
  var hostParts = (match || []).slice(1, 3)

  // ES's hostParts are sometimes the other way round, if the value that is expected
  // to be region equals ‘es’ switch them back
  // e.g. search-cluster-name-aaaa00aaaa0aaa0aaaaaaa0aaa.us-east-1.es.amazonaws.com
  if (hostParts[1] === 'es')
    hostParts = hostParts.reverse()

  return hostParts
}
```

**Comment by mhart** [comment]
(it's cool, bit late in London :smile:  – I'll restructure this)

Thanks again!

**Comment by mhart** [comment]
Published as v1.1.0

**Comment by matthew-andrews** [comment]
Thanks! :tada:

**Pull request merged**